PEP 767: Annotating Read-Only Attributes (#4127)
python/peps: peps/pep-0767.rst

PEP: 767
Title: Annotating Read-Only Attributes
Author: Eneg <eneg at discuss.python.org>
Sponsor: Carl Meyer <carl@oddbird.net>
Discussions-To: <REQUIRED: URL of current canonical discussion thread>
Status: Draft
Type: Standards Track
Topic: Typing
Created: 18-Nov-2024
Python-Version: 3.14

Abstract
========

:pep:`705` introduced the :external+py3.13:data:`typing.ReadOnly` type qualifier
to allow defining read-only :class:`typing.TypedDict` items.

This PEP proposes expanding the scope of ``ReadOnly`` to class and protocol

> attributes, as a single concise way to mark them read-only. Some parity changes
> are also made to :data:`typing.Final`.

Reviewer: In the current draft the changes to `Final` are limited to clarifying that final attributes cannot be deleted, right? Maybe rephrase this to better reflect that the situation, something like
```suggestion
attributes, as a single concise way to mark them read-only. One small clarification
is also made to the specification of :data:`typing.Final`.
```

Author: I was planning to remove this note altogether, but this is fine too

> Akin to the :pep:`705`, it makes no changes to setting attributes at runtime. Correct

Reviewer: ```suggestion
Akin to :pep:`705`, it makes no changes to setting attributes at runtime. Correct
```

Author: English hard 😄
I thought "the" is correct here given the PEP is already mentioned two paragraphs above

Reviewer: It would be correct to say "Akin to the PEP" if the referent of "the PEP" were clear in context without any disambiguation, but that is not the case here, despite it being previously mentioned. ("The PEP" could easily mean the current PEP.)

If we are saying "PEP 705" (which is what the Sphinx role will expand to), then we don't refer to individual PEPs with the definite article; we say "PEP 705", not "the PEP 705"; whether it's been referred to previously doesn't change this.

to forbid mutation of such values:

.. code-block:: python

    from collections import abc
    from dataclasses import dataclass
    from typing import Protocol, ReadOnly

    @dataclass
    class Game:
        name: str

    class HasGames[T: abc.Collection[Game]](Protocol):
        games: ReadOnly[T]

    def add_games(shelf: HasGames[list[Game]]) -> None:
        shelf.games.append(Game("Half-Life"))  # ok: list is mutable
        shelf.games[-1].name = "Black Mesa"    # ok: "name" is not read-only
        shelf.games = []                       # error: "games" is read-only
        del shelf.games                        # error: "games" is read-only and cannot be deleted

    def read_games(shelf: HasGames[abc.Sequence[Game]]) -> None:
        shelf.games.append(...)             # error: "Sequence" has no attribute "append"
        shelf.games[0].name = "Blue Shift"  # ok: "name" is not read-only
        shelf.games = []                    # error: "games" is read-only

All instance attributes of frozen dataclasses and ``NamedTuple`` should be
implied to be read-only. Type checkers **should not** flag redundant annotations

> of such attributes with ``ReadOnly``:

Reviewer: I wonder if it's important that type checkers not be allowed to flag this? Personally I would find the redundant `ReadOnly` annotation below to be marginally bad style (particularly that some attributes are explicitly marked and others not) and would welcome a type checker offering an "info" or "warning" message that the `ReadOnly` annotation is redundant here and can be removed. So I might consider specifying instead that a redundant annotation is not an error, but type checkers may optionally choose to warn about it.

PEP: 767
Title: Read-only class and protocol attributes
Author: <REQUIRED: list of authors' real names and optionally, email addrs>
Sponsor: Carl Meyer <carl@oddbird.net>
Discussions-To: <REQUIRED: URL of current canonical discussion thread>
Status: Draft
Type: Standards Track
Topic: Typing
Created: 18-Nov-2024
Python-Version: 3.14

Abstract
========

:pep:`705` introduced the :external+py3.13:data:`typing.ReadOnly` type qualifier
to allow defining read-only :class:`typing.TypedDict` items.

This PEP proposes expanding the scope of ``ReadOnly`` to class and protocol
attributes, as a single concise way to mark them read-only. Some parity changes
are also made to :data:`typing.Final`.

Akin to the :pep:`705`, it makes no changes to setting attributes at runtime. Correct
usage of read-only attributes is intended to be enforced only by static type checkers.

Motivation
==========

The Python type system lacks a single concise way to mark an :term:`attribute` read-only.

> This feature is present in other statically and gradually typed languages

Author: I've changed "object-oriented" to "statically and gradually typed", hopefully it's not a misnomer

    @dataclass(frozen=True)
    class UnsignedPoint(Point):
        x: ReadOnly[uint]  # ok, explicit & narrower type
        y: uint            # ok, narrower type

Initialization
--------------

Assignment to a read-only attribute can only occur in the class declaring the attribute.
There is no restriction to how many times the attribute can be assigned to.
The assignment must be allowed in the following contexts:

1. In ``__init__``, on the instance received as the first parameter (likely, ``self``).
2. In ``__new__``, on instances of the declaring class created via a call
   to a super-class' ``__new__`` method.
3. At declaration in the body of the class.

Additionally, a type checker may choose to allow the assignment:

1. In ``__new__``, on instances of the declaring class, without regard
   to the origin of the instance.
   (This choice trades soundness, as the instance may already be initialized,
   for the simplicity of implementation.)
2. In ``@classmethod``\ s, on instances of the declaring class created via
   a call to the class' or super-class' ``__new__`` method.

Note that a child class cannot assign to any read-only attribute of a parent class

> in any of the aforementioned contexts, unless the attribute is redeclared.

Author: @carljm how do you like this?

Reviewer: I like it! We'll see if other reviewers agree :)

The code examples below are very clarifying for the other rules; I think it would be useful to expand them with examples of `__new__` and classmethods.

It may also be worth a section in "rejected alternatives" discussing _why_ being able to assign readonly attributes in classmethod constructors is useful (that is, the `Fraction` example.)

Author: What exactly would be the "rejected alternative"? Disallowing classmethods/assignment only in `__init__`?

Reviewer: Yes, and the text would summarize the reasoning for allowing assignment in `__new__` and optionally other class methods.

        def __init__(self) -> None:
            self.songs = []  # error: cannot assign to a read-only attribute of a base class

.. code-block:: python

    # a simplified immutable Fraction class
    class Fraction:
        numerator: ReadOnly[int]
        denominator: ReadOnly[int]

        def __new__(
            cls,
            numerator: str | int | float | Decimal | Rational = 0,
            denominator: int | Rational | None = None
        ) -> Self:
            self = super().__new__(cls)

            if denominator is None:
                if type(numerator) is int:
                    self.numerator = numerator
                    self.denominator = 1
                    return self

                elif isinstance(numerator, Rational): ...

            else: ...

        @classmethod
        def from_float(cls, f: float, /) -> Self:
            self = super().__new__(cls)
            self.numerator, self.denominator = f.as_integer_ratio()
            return self

When a class-level declaration has an initializing value, it can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
default for instances:

.. code-block:: python

    class Patient:

>         number: ReadOnly[int] = 0

Reviewer 2: Are the semantics of omitting the type explained? Example:
```py
    class Patient:
        number: ReadOnly = 0  # Is this ok? What's the type of number?
```
If type inference is supported, it would be useful to have some guidelines about how type checkers should approach this, even if some details are omitted (similar to other PEPs that don't specify type inference in detail).

Reviewer 3: There was some discussion on this here: https://github.com/python/peps/pull/4127#discussion_r1849261608

    class MyBase(abc.ABC):
        foo: ReadOnly[int]
        bar: ReadOnly[str] = "abc"
        baz: ReadOnly[float]

        def __init__(self, baz: float) -> None:
            self.baz = baz

        @abstractmethod
        def pprint(self) -> None: ...

    @final
    class MySubclass(MyBase):
        # error: MySubclass does not override "foo"

        def pprint(self) -> None:
            print(self.foo, self.bar, self.baz)

* In a protocol attribute declaration, ``name: ReadOnly[T]`` indicates that a structural
  subtype must support ``.name`` access, and the returned value is assignable to ``T``::

    class HasName(Protocol):
        name: ReadOnly[str]

    class NamedAttr:
        name: str

    class NamedProp:
        @property
        def name(self) -> str: ...

    class NamedClassVar:
        name: ClassVar[str]

    class NamedDescriptor:
        @cached_property
        def name(self) -> str: ...

    # all of the following are ok
    has_name: HasName
    has_name = NamedAttr()
    has_name = NamedProp()
    has_name = NamedClassVar
    has_name = NamedClassVar()
    has_name = NamedDescriptor()

> Interaction with Other Special Types
> ------------------------------------

Reviewer 4: ```suggestion
Interaction with Other Type Qualifiers
--------------------------------------
```
ClassVar etc. are not types, but type qualifiers
